File: problem328.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode OddEvenList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode evenHead = head.next;
        ListNode currOdd = head;
        ListNode currEven = head.next;
        while (currOdd != null && currEven != null && currOdd.next != null && currEven.next != null) {
            currOdd.next = currEven.next;
            currOdd = currOdd.next;
            currEven.next = currOdd.next;
            currEven = currEven.next;
        }
        currOdd.next = evenHead;
        return head;
    }
}

File: su19/problem328.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None:
            return None
        oddStart = odds = head
        evenStart = evens = head.next
        while odds != None and evens != None:
            odds.next = evens.next
            if odds.next is not None:
                odds = odds.next
            evens.next = odds.next
            evens = evens.next
            
        odds.next = evenStart
        return oddStart